File: F/833-find-and-replace-in-string/main.go
package main

import "fmt"

func findReplaceString(S string, indexes []int, sources []string, targets []string) string {
	buf := []byte(S)
	M := map[int]int{}
	for i := 0; i < len(indexes); i++ {
		idx := indexes[i]
		isMatch := true
		for j := 0; j < len(sources[i]); j++ {
			if buf[idx+j] != sources[i][j] {
				isMatch = false
				break
			}
		}
		if isMatch {
			for j := 0; j < len(sources[i]); j++ {
				buf[idx+j] = '0'
			}
			M[idx] = i
		}
	}

	var ret []byte
	idx := 0
	for idx < len(buf) {
		if buf[idx] <= 'z' && buf[idx] >= 'a' {
			ret = append(ret, buf[idx])
			idx++
			continue
		}

		tIdx := M[idx]
		for i := 0; i < len(targets[tIdx]); i++ {
			ret = append(ret, targets[tIdx][i])
		}

		i := 0
		for idx < len(buf) && buf[idx] == '0' && i < len(sources[tIdx]) {
			idx++
			i++
		}
	}
	return string(ret)
}

func main() {
	fmt.Println()
}

File: F/833-find-and-replace-in-string/main.rs
struct Solution {}

impl Solution {
  pub fn find_replace_string(
    s: String,
    indices: Vec<i32>,
    sources: Vec<String>,
    targets: Vec<String>,
  ) -> String {
    let mut bb: Vec<u8> = s.as_bytes().to_vec();
    let mut replace: Vec<i32> = vec![0; indices.len()];
    (0..indices.len()).for_each(|i| {
      let mut offset = indices[i] as usize;
      let mut end = offset + sources[i].len();
      while offset < bb.len()
        && offset < end
        && bb[offset] == sources[i].as_bytes()[offset - indices[i] as usize]
      {
        offset += 1;
      }
      if offset == end {
        replace[i] = 1;
        bb[indices[i] as usize] = 0;
      }
    });
    let mut res: Vec<u8> = vec![];
    let mut idx: usize = 0;
    while idx < bb.len() {
      if bb[idx] == 0 {
        let i = indices.iter().position(|&x| x as usize == idx).unwrap();
        res.extend(targets[i].as_bytes());
        idx += sources[i].len();
      } else {
        res.push(bb[idx]);
        idx += 1;
      }
    }
    String::from_utf8(res).unwrap()
  }
}
